pkg/docker/types/types_easyjson.go

```diff
@@ -65,7 +62,7 @@ func easyjson6601e8cdDecodeGithubComStackroxRoxPkgDockerTypes(in *jlexer.Lexer,
 				}
 				for !in.IsDelim(']') {
 					var v1 ContainerJSON
-					easyjson6601e8cdDecodeGithubComStackroxRoxPkgDockerTypes1(in, &v1)
+					(v1).UnmarshalEasyJSON(in)
```

Comment:
Do you know, why is there such a large diff on `types_easyjson.go`?

Reply:
Yeah, this was a fun puzzle to solve while working on this: I observed that the generation apparently was unstable, i.e., I ran `make go-easyjson-srcs`, checked in the code, and saw the "check that generated sources are up to date" check fail. At first I thought there must be some version mismatch. Then I ran the whole `make go-generated-srcs` (which I almost never run locally, and also didn't in this PR before), and voila, I got matching easyjson sources again. Then I re-ran `make go-easyjson-srcs`, and it broke.

I then took a closer look and saw that `make go-generated-srcs` depends on `clean-easyjson-srcs`, and that `go-easyjson-srcs` generates the "right" (wrt. CI) sources whenever the easyjson srcs don't exist. I looked up the `easyjson` help, and saw the help text
```
  -pkg
        process the whole package instead of just the given file
```
so I became suspicious that all files are listed individually. Apparently, "process" means "generate optimized marshal/unmarshal code", but the exported methods will still only be generated for the annotated types specified in the `.go` file. However, `-pkg` also causes `easyjson` to pick up any optimized marshaling/unmarshaling code that's already there in other files for some of the used types (but if it doesn't find optimized methods, it will generate marshaling/unmarshaling for types outside of the `.go` file with unexported names). Running `make go-easyjson-srcs` when the sources were already there just picked up those types/methods that weren't picked up by it running in CI in the course of `make go-generated-srcs`, that cleaned them beforehand. So I changed the generation order to process files in topological order, making sure that easyjson generation for types in one file has always finished by the time the file in which those types are used is processed.

This is ugly, to say the least, and extremely brittle - especially because you absolutely can have circular dependencies between `.go` files, and thus you can break this in a way that's not only hard but impossible to repair (well, I guess one could always run all generation commands twice). But other than put a warning in the `Makefile` to honor the order, I don't know a better way to fix it. And before you ask, no - specifying `-pkg` and then _all_ files in the directory doesn't work, either. This would probably be a good change in `easyjson`, but unfortunately, I [didn't have any luck with PRs in the past](https://github.com/mailru/easyjson/pull/376).

So yeah, the diff is IMHO what it should be, without the duplicate generated code that was there before. It's also stable now, in the sense that running `make go-easyjson-srcs` always gives the same result, regardless of whether easyjson files are there or not. But I'm not at all happy at how I've reached that state.

Reply:
Thanks for the explanation. I find `-pkg` to be very confusing. I would expect that it can be sufficient to run `$(SILENT)$(EASYJSON_BIN) -pkg pkg/docker/types/<blah>.go` only for **one file** in that directory and it will pick up two others as they have matching `package types`, but that doesn't happen.

Moreover, if I take your branch and remove `-pkg` from arguments like this

```makefile
.PHONY: go-easyjson-srcs
go-easyjson-srcs: $(EASYJSON_BIN)
	@echo "+ $@"
	@# Don't change the order of these, as otherwise the generation might become unstable.
	$(SILENT)$(EASYJSON_BIN) pkg/docker/types/image.go
	$(SILENT)$(EASYJSON_BIN) pkg/docker/types/container.go
	$(SILENT)$(EASYJSON_BIN) pkg/docker/types/types.go
	$(SILENT)$(EASYJSON_BIN) pkg/compliance/compress/compress.go
```

I see the diff is empty after running `make clean-easyjson-srcs` and `make go-easyjson-srcs` in any combinations.

Looks like `-pkg` doesn't have any effect for us. Therefore I suggest dropping it to avoid misleading ourselves.

Reply:
I agree with your expectation on what `-pkg` does, and the frustration about this not being the reality. The fact that it isn't even needed in our case leaves me completely puzzled as to what its point is. Will remove.